astropy/stats/spatial.py

```diff
@@ -0,0 +1,314 @@
+
+        Parameters
+        ----------
+        data : 2D array
+            Set of observed points in as a n by 2 array which will be used to
+            estimate Ripley's K function.
+        radii : 1D array
+            Set of distances in which Ripley's K estimator will be evaluated.
+            Usually, it's common to consider max(radii) < (area/2)**0.5.
+        mode : str
+            Keyword which indicates the method for edge effects correction.
+            Available methods are 'none', 'translation', 'ohser', 'var-width',
+            and 'ripley'.
+
+            * 'none'
+                this method does not take into account any edge effects
+                whatsoever.
+            * 'translation'
+                computes the intersection of rectangular areas centered at
+                the given points provided the upper bounds of the
+                dimensions of the rectangular area of study. It assumes that
+                all the points lie in a bounded rectangular region satisfying
+                x_min < x_i < x_max; y_min < y_i < y_max. A detailed
+                description of this method can be found on ref [4].
+            * 'ohser'
+                this method uses the isotropized set covariance function of
+                the window of study as a weigth to correct for
+                edge-effects. A detailed description of this method can be
+                found on ref [4].
+            * 'var-width'
+                this method considers the distance of each observed point to
+                the nearest boundary of the study window as a factor to
+                account for edge-effects. See [3] for a brief description of
+                this method.
+            * 'ripley'
+                this method is known as Ripley's edge-corrected estimator.
+                The weight for edge-correction is a function of the
+                proportions of circumferences centered at each data point
+                which crosses another data point of interest. See [3] for
+                a detailed description of this method.
+
+        Returns
+        -------
+        ripley : 1D array
+            Ripley's K function estimator evaluated at ``radii``.
+        """
+
+        data = np.asarray(data)
+
+        if not data.shape[1] == 2:
+            raise ValueError('data must be an n by 2 array, where n is the '
+                             'number of observed points.')
+
+        npts = len(data)
+        ripley = np.zeros(len(radii))
+
+        if mode == 'none':
+            distances = self._pairwise_distances(data)
+            for r in range(len(radii)):
+                ripley[r] = (distances < radii[r]).sum()
+
+            ripley = self.area * 2. * ripley / (npts * (npts - 1))
+        # eq. 15.11 Stoyan book page 283
+        elif mode == 'translation':
+            diff = self._pairwise_diffs(data)
+            distances = self._pairwise_distances(data)
```

Comment:
That seems like an unnecessary recalculation because `self._pairwise_distances` already calculates the `diffs` so these are calculated two times. But I have currently no suggestion how to improve that. :)